internal/provider/common_cert.go

```diff
@@ -0,0 +1,463 @@
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"math/big"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+var keyUsages = map[string]x509.KeyUsage{
+	"digital_signature":  x509.KeyUsageDigitalSignature,
+	"content_commitment": x509.KeyUsageContentCommitment,
+	"key_encipherment":   x509.KeyUsageKeyEncipherment,
+	"data_encipherment":  x509.KeyUsageDataEncipherment,
+	"key_agreement":      x509.KeyUsageKeyAgreement,
+	"cert_signing":       x509.KeyUsageCertSign,
+	"crl_signing":        x509.KeyUsageCRLSign,
+	"encipher_only":      x509.KeyUsageEncipherOnly,
+	"decipher_only":      x509.KeyUsageDecipherOnly,
+}
+
+var extendedKeyUsages = map[string]x509.ExtKeyUsage{
+	"any_extended":                      x509.ExtKeyUsageAny,
+	"server_auth":                       x509.ExtKeyUsageServerAuth,
+	"client_auth":                       x509.ExtKeyUsageClientAuth,
+	"code_signing":                      x509.ExtKeyUsageCodeSigning,
+	"email_protection":                  x509.ExtKeyUsageEmailProtection,
+	"ipsec_end_system":                  x509.ExtKeyUsageIPSECEndSystem,
+	"ipsec_tunnel":                      x509.ExtKeyUsageIPSECTunnel,
+	"ipsec_user":                        x509.ExtKeyUsageIPSECUser,
+	"timestamping":                      x509.ExtKeyUsageTimeStamping,
+	"ocsp_signing":                      x509.ExtKeyUsageOCSPSigning,
+	"microsoft_server_gated_crypto":     x509.ExtKeyUsageMicrosoftServerGatedCrypto,
+	"netscape_server_gated_crypto":      x509.ExtKeyUsageNetscapeServerGatedCrypto,
+	"microsoft_commercial_code_signing": x509.ExtKeyUsageMicrosoftCommercialCodeSigning,
+	"microsoft_kernel_code_signing":     x509.ExtKeyUsageMicrosoftKernelCodeSigning,
+}
+
+// supportedKeyUsages returns a slice with all the keys in keyUsages and extendedKeyUsages.
+func supportedKeyUsages() []string {
+	res := make([]string, 0, len(keyUsages)+len(extendedKeyUsages))
+
+	for k := range keyUsages {
+		res = append(res, k)
+	}
+	for k := range extendedKeyUsages {
+		res = append(res, k)
+	}
+	sort.Strings(res)
+
+	return res
+}
+
+// generateSubjectKeyID generates a SHA-1 hash of the subject public key.
+func generateSubjectKeyID(pub crypto.PublicKey) ([]byte, error) {
+	var publicKeyBytes []byte
+	var err error
+
+	switch pub := pub.(type) {
+	case *rsa.PublicKey:
+		publicKeyBytes, err = asn1.Marshal(*pub)
+		if err != nil {
+			return nil, err
+		}
+	case *ecdsa.PublicKey:
+		publicKeyBytes = elliptic.Marshal(pub.Curve, pub.X, pub.Y)
+	default:
+		return nil, errors.New("only RSA and ECDSA public keys supported")
+	}
+
+	hash := sha1.Sum(publicKeyBytes)
+	return hash[:], nil
+}
+
+// setCertificateSubjectSchema sets on the given reference to map of schema.Schema
+// all the keys required by a resource representing a certificate's subject.
+func setCertificateSubjectSchema(s *map[string]*schema.Schema) {
+	(*s)["dns_names"] = &schema.Schema{
```

Comment:
Nit: Using map pointers is extraneous -- these can be simplified to 

```go
func setCertificateSubjectSchema(s map[string]*schema.Schema) {
	s["dns_names"] = &schema.Schema{
```

etc. 👍

Reply:
Yeah, I forgot to integrate that feedback. Will do now.